docker/src/main/java/clocker/docker/location/DockerLocation.java

```diff
@@ -245,13 +239,20 @@ public MachineLocation obtain(Map<?,?> flags) throws NoMachinesAvailableExceptio
         Entities.waitForServiceUp(dockerHost);
 
         // Obtain a new Docker container location, save and return it
-        if (LOG.isDebugEnabled()) {
+        try {
             LOG.debug("Obtain a new container from {} for {}", machine, entity);
+            Map<?,?> hostFlags = MutableMap.copyOf(flags);
+            DockerContainerLocation container = machine.obtain(hostFlags);
+            containers.put(machine, container.getId());
+            return container;
+        } finally {
+            // Release any placement strategy locks
+            for (DockerAwarePlacementStrategy strategy : Iterables.concat(strategies, entityStrategies)) {
+                if (strategy instanceof WithMutexes) {
+                    ((WithMutexes) strategy).releaseMutex(entity.getApplicationId());
```

Comment:
This try-finally block in `DockerLocation` is surprising. After calling `strategy.filterLocations`, there is a big chunk of code not guarded by the finally block, so if there is an exception during that work, the acquired mutex will not be released.